### ISSUE TITLE
Scroll through every object in a class using an id cursor

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#2302]
The idea is simple: Using pagination you can at most retrieve `QUERY_MAXIMUM_RESULTS` objects. To retrieve more objects you would need to start "after" the last ID you retrieved. There is currently no way to do this.

Note: This concept relies on the order of ids and can therefore only be applied to list queries that are unfiltered and don't contain any sorting. Otherwise the same problem as with pagination would occur. So this would only solve the use case of getting all objects of a class.

## API
Cursor works similar to pagination, but instead of an offset, an `after` flag can be defined. `after=null` is equivalent to `offset=0`. Setting both `after` and `offset` at the same time is not permitted. The user can use the last id from the previous request to get the next page. `After` always takes a uuid.

### REST
Retrieve 10 objects starting at id `1b412672-ba9c-4804-bce0-c33d77f11558` of class `Article`
```http
GET /v1/objects?class=Article&after=1b412672-ba9c-4804-bce0-c33d77f11558&limit=10
```
Notes:
*  If `class` is not set, after cannot be set. A cursor is always specific to one class`

### GraphQL
Retrieve 10 objects starting at id `1b412672-ba9c-4804-bce0-c33d77f11558` of class `Article`:

```graphql
{
  Get {
    Article(
      limit:10
      after: "1b412672-ba9c-4804-bce0-c33d77f11558"
    ){
      prop_1
    }
  }
}
```
Notes:
* `after` is only permitted if no search operator is set. `after` is not compatible with `where`, `near<Media>`, `bm25`, `hybrid`, etc. For those cases, the user can use `offset`

## Usage Examples

The most popular use case for the cursor API is to get all objects out of Weaviate. An example in pseudo-code could look like the following:

```python
last_id=None
results=[]
while True:
  batch = client.get(class_name="Article", limit=100, after=last_id)
  if len(batch) == 0:
    break

  results.append(batch)
  last_id=batch[len(batch)-1].id
```

The above implies that:
* When the user starts with a `None`/`null` id in the first round, the cursor returns the first object
* When the cursor API is called with the last id, it returns 0 results, the client uses this as the exit condition

### Single-Shard
For a single-shard setup, this is very straightforward. It's essentially just using the `Cursor()` API on the `objects` `lsmkv.Bucket`.

### Multi-shard
For multi-shard, we need to aggregate the results and then re-sort by UUID and cut off at the desired limit. This means that we are over-fetching a bit for each request on a multi-shard setup. This is fine, the main motivation behind the cursor API at the moment is to give users a way to get all objects of a class, not on efficiency. This over-fetching pattern would be similar to a multi-shard search for example.

**Comment by etiennedi** [comment]
Thank you all for the votes! This is the clear "winner" of the backlog. Let's tackle this in `v1.18`! 🎉

**Comment by etiennedi** [comment]
Not stale, this is already planned.

**Comment by trengrj** [comment]
When we release this feature we should also add an `export` function to the clients that would use the scroll api to extract all files, vectors, ids, out to a dictionary.

**Comment by etiennedi** [comment]
I like the idea of an export function, but I'm not sure that we should tie it to this feature. We could also release the export function later, otherwise, this could risk our timelines for v1.18. Let's create a separate ticket for it.

**Comment by etiennedi** [comment]
The OP is updated with the API proposal.

@trengrj, @byronvoorbach Any thoughts on the name of the parameter `after`? Is it clear that it refers to an id or would you prefer `after_id`/`afterId`? Or a completely different name?

**Comment by dirkkul** [comment]
this is done and released in v1.18
